CI: add dev-only Windows smoke workflow

kad-link opened the pull request:
## Description
This PR adds an experimental Windows CI workflow to provide a lightweight signal for Windows contributors. 

- Adds `windows-ci.yml` running on `windows-latest` for push and pull request
- Runs `flake8` on Windows (same lint target as Linux CI)
- Runs `python -m compileall` as a smoke check to catch syntax or import-time issues
- Marked as non-blocking via `continue-on-error: true`

### Note
This workflow is intentionally dev-only and does not attempt a full nominatim-db install on Windows (blocked by PyICU/ICU native build requirements on GitHub-hosted runners). Linux CI remains the authoritative test pipeline.

## Future Work
This is the first step toward full Windows support. I am planning a follow-up effort to resolve the PyICU / ICU native build issues so we can eventually run full Nominatim installs and tests on Windows CI.

kad-link commented:
> That's a nice start but we'll need a full run of Nominatim's tests to make sure that the installation actually works. At a minimum the [tests workflow](https://github.com/osm-search/Nominatim/blob/20a333dd9b3649699b2e5269cda95c4f30b00aae/.github/workflows/ci-tests.yml#L36) should work just like on the Linux installations. Ideally, we get a [install workflow](https://github.com/osm-search/Nominatim/blob/20a333dd9b3649699b2e5269cda95c4f30b00aae/.github/workflows/ci-tests.yml#L143) to work as well.

Thanks for the feedback. 
To spread this further in an organized way, I would like to approach this in stages (with one PR per stage) to keep things maintainable and reviewable. 
### Stage 1 : (Windows Test Parity) 
- Setup PostgreSQL on Windows
- Build the required Python environment
- Run the same Python unit tests and BDD tests that currently run on Linux (or as close as possible, skipping only where Windows-specific blockers exist)

### Stage 2 : 
Add full import support + osm2pgsql so that BDD import scenarios run successfully.

### Stage 3 : 
Replicate the full install workflow on Windows to match Linux install validation.

I’ll start with Stage 1 so we can first get reliable test coverage on Windows and iterate from there.
Please let me know if this plan aligns with expectations.

kad-link commented:
> Restricting CI tests to the Python unit tests in a first stage PR sounds reasonable. Lets start with that.

Hi @lonvia , while going through the current [ci-tests.yml](https://github.com/osm-search/Nominatim/blob/master/.github/workflows/ci-tests.yml) , I noticed the `create-archive` job which prepares the source tarball and later consumed by subsequent jobs (including Python unit-tests). 
Since I am working on adding CI coverage for Windows, I am wondering if it should be implemented in the CI for Windows too. Since `create-archive` is only packaging the source tree , I think it would be reasonable to keep this job running on Linux, while the other jobs run on Windows. Linux runners seem well suited for shell-based packaging steps.

So, could you kindly clarify these 2 issues in summary ?
-  Should the Windows CI workflow include the `create-archive` job?
- If so, is it preferred to keep  the `create-archive` job Linux-only (while the rest of the code Windows-specific), or should there be a Windows-specific implementation for the `create-archive` job too?

kad-link commented:
Hi @mtmail , I have resolved the above problem but I'm seeing mypy failures on `src/nominatim_api/server/starlette/server.py`(`request.state.num_results` and `request.state.query_stats` attr-defined). Now what's confusing is the exact same mypy step (`mypy --strict --python-version 3.9 src`) passed in the previous CI run and I didn't modify this file. Could you confirm if this is expected or if we want to resolve this newly risen issue in the same PR?

mtmail commented:
@lonvia knows mypy better, there always seem to be tiny changes between versions. In the meantime I would skip the test or make that section optional and concentrate on the unit/bdd tests.

mtmail commented:
@mtmail Let's keep this PR open, then all Windows-test related discussion is in one place.

Can you create a new PR for the changes to src/nominatim_api/server/starlette/server.py ? In every release mypy keeps adding more warnings and this one is the same for all operating systems. You found a good solution and I'd be good if a PR with your name attached gets merged instead of me repeating the changes. Something like "fixes a new mypy type warning related to Starlette state object"

lonvia commented:
The main issue with Postgres on Windows used to be that it doesn't come with Postgis. If that's still the case, it might help to have a look at the [action script from osm2pgsql](https://github.com/osm2pgsql-dev/osm2pgsql/blob/master/.github/actions/win-postgres/action.yml) which downloads and installs postgis. Not sure it is still the best way of doing things, but it is one that works.

kad-link commented:
Also, lemme tell you about the fix , 
I identified that this can be permanently resolved using `pytest-xdist` grouping. The following is the implementation : 
- Add a marker in `conftest.py` to group the API tests 
- Update the CI workflow to run with `--dist=loadgroup`


This forces the conflicting tests to run on the same worker sequentially while keeping the rest of the tests running parallel. 
A glimpse of the marker goes like this : 
`item.add_marker(pytest.mark.xdist_group(name = "api_exclusive"))`

lonvia commented:
None of the tests should be run in parallel. We use hard-coded database names and the tests would get into each others way. Where have you see pytest-xdist used in the CI?

Note: there is an occasional fail on deadlock in the tests during installation of the functions. That's an actual issue with the code, not a problem of the tests. I just haven't been able to pin it down yet because it is difficult to reproduce.

lonvia commented:
Looks all good so far. Just keep going to make the Python tests pass.

pyICU is a bit of a difficult one because they don't provide wheels with the shared library included. They expect libicu to be present externally. I see that somewhat has made full packages for Windows [here](https://github.com/cgohlke/pyicu-build/releases). Maybe try those. That might be easier than installing libicu on the Windows system.

kad-link commented:
Today's CI failures resolved successfully : 


- `UnicodeDecodeError` (Encoding):
   Issue: Windows runners default to `CP1252` encoding, causing `refresh.py` to crash when reading UTF8 files.
    Fix: Implemented `PYTHONUTF8=1` in the workflow environment to ensure uniform file handling.

- `OperationalError` (Spatialite issues):
    Issue: SQLite tests were failing because `mod_spatialite` and its dependencies (e.g., `libgeos.dll`) could not be found.
    Fix: Added a step to download `mod_spatialite-5.1.0-win-amd64` and correctly added its directory to `$env:GITHUB_PATH` so that Python can load it.

- `WinError 2` (Missing Executables):
    Issue: `osm2pgsql` related tests were crashing because the corresponding binaries are not yet present in the Windows CI. 
    Fix: Added `--ignore` flags to the `pytest` command to skip its tests

kad-link commented:
> > ```
> > * `UnicodeDecodeError` (Encoding):
> >   Issue: Windows runners default to `CP1252` encoding, causing `refresh.py` to crash when reading UTF8 files.
> >   Fix: Implemented `PYTHONUTF8=1` in the workflow environment to ensure uniform file handling.
> > ```
> 
> Isn't that just a band aid solution? If we expect things to be UTF8, we should specify that encoding when we're reading files.



Yes, agreed that it's a temporary patch to avoid CI errors. I’ll manually add `encoding='utf-8'` in the relevant file reads.

lonvia commented:
> I think modifying many files in this PR will make it messy and chaotic. So can we keep PYTHONUTF8=1 for now to get the Windows CI going? I can maybe open a separate follow up PR dedicated to fixing the file encodings.

Sounds reasonable. Maybe open an issue as a reminder that this needs looking into?

kad-link commented:
> > I think modifying many files in this PR will make it messy and chaotic. So can we keep PYTHONUTF8=1 for now to get the Windows CI going? I can maybe open a separate follow up PR dedicated to fixing the file encodings.
> 
> Sounds reasonable. Maybe open an issue as a reminder that this needs looking into?

I will also open up this issue shortly

lonvia commented:
One more thing. There is currently a deprecation warning: 

```
test\python\conftest.py:13
  D:\a\Nominatim\Nominatim\Nominatim\test\python\conftest.py:13: DeprecationWarning: 'asyncio.WindowsSelectorEventLoopPolicy' is deprecated and slated for removal in Python 3.16
    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
```

Can you please find the appropriate replacement?

kad-link commented:
hi @lonvia , I have removed the packaging as discussed. The tests have passed successfully. My next checkpoints are to upgrade the PostgreSQL version to 18 and find a substitute for this deprecation warning : 

> ```
> test\python\conftest.py:13
>   D:\a\Nominatim\Nominatim\Nominatim\test\python\conftest.py:13: DeprecationWarning: 'asyncio.WindowsSelectorEventLoopPolicy' is deprecated and slated for removal in Python 3.16
>     asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
> ```

kad-link commented:
Done @lonvia

lonvia commented:
Alright, lets see how this goes with daily use. Thanks a lot for patiently keeping on this.

Would you mind opening issues for the remaining todos? Preferably a separate one for each.
